velox/substrait/VeloxToSubstraitExpr.cpp

```diff
@@ -465,7 +466,41 @@ VeloxToSubstraitExprConvertor::toSubstraitExpr(
   ::substrait::Expression_ReferenceSegment_StructField* directStruct =
       substraitFieldExpr->mutable_direct_reference()->mutable_struct_field();
 
-  directStruct->set_field(inputType->getChildIdx(exprName));
+  // FieldAccessTypedExpr represents one of two things: a leaf in an expression
+  // or a dereference expression(fieldExpr->isInputColumn() == false)
+  // for a leaf in an expression, find idx from child by exprName.
+  // for a dereference expression, find idx from every child by exprName.
+  if (fieldExpr->isInputColumn() ||
```

Comment:
I don't understand this code and the API change to pass currentInputType looks strange. Would you explain how this code works? Can we avoid API change?

Reply:
Thanks for your review.  I indeed want to avoid API change. 
But I found the current variable can not achieve my goal, when I try to use recursive way to constuct StructField.
For function toSubstraitExpr. inputType is a const shared_ptr. So the address of itself can not be changed. 

For expression like (abc).ab.a.
fieldExpr is < a<ab <abc [input]>>>. 'a' is the outermost FieldAccessTypedExpr.
But inputType is "ROW<abc:ROW<ab:ROW<a:BIGINT,b:BIGINT>,c:BIGINT>>".  'a' is the innermost type.

So I need a shared_pointer(currentInputType) to tell function toSubstraitExpr where are we now.
The simple logic for toSubstraitExpr is shown as following:
1. reach the innermost layer of  fieldExpr(abc). 
2. set set_field for the outermost struct. 
3. set currentInputType point to the child of corresponding FieldAccessTypedExpr.
4. end recusion, when we arrive the outermost layer of fieldExpr(a).

Here is a picture of call process.
![image](https://user-images.githubusercontent.com/14065538/219852645-cbb1557c-f3e9-47f5-bb35-5fe2679f605f.png)

Reply:
@zhejiangxiaomai Let's look at this slightly differently. An expression is a tree with different kinds of nodes. We have constant, fieldAccess, cast and call nodes. All these nodes can be combined in any way. 'inputType' is the schema of the input to the whole expression. Leaf nodes directly access data of that schema. Intermediate nodes access the data produces by their inputs. When converting fieldAccess expression we need to distinguish the case where it is a leaf node or intermediate node. When fieldAccess is a leaf node the "name" of the field refers to a column in 'inputType'. When fieldAccess is an intermediate node the "name" refers to a column in the output of the input node. To convert intermediate field access we need to (1) convert the input, (2) lookup the "name" in the output of the "input". Hope this helps.